content/posts/using-the-redis-allocator-in-rust.md

```diff
@@ -0,0 +1,153 @@
+---
+title: "Using the Redis Allocator in Rust"
+date: 2019-11-12
+authors:
+  - Gavrie Philipson
+tags: ["redisjson", "rust"]
+---
+
+## Introduction
+
+While developing [redismodule-rs](https://github.com/RedisLabsModules/redismodule-rs), the [Rust](https://www.rust-lang.org) API for writing [Redis modules](https://redis.io/topics/modules-intro), I encountered the need to set up a custom memory allocator.
+
+Normally, when a Rust program needs to allocate some memory, such as when creating a `String` or `Vec` instance, it uses the [global allocator](https://doc.rust-lang.org/std/alloc/index.html) defined in the program. Since Redis modules are built as shared libraries to be loaded into Redis, Rust will use the [`System`](https://doc.rust-lang.org/std/alloc/struct.System.html) allocator, which is the default provided by the OS (using the `libc` [`malloc(3)`](https://linux.die.net/man/3/malloc) function).
+
+This behavior is problematic for several reasons. 
+
+First of all, Redis may not be using the system allocator at all, relying on [`jemalloc`](http://jemalloc.net) instead. The `jemalloc` allocator is an alternative to the system `malloc` that includes many tweaks to avoid fragmentation, among other features. If the module uses the system allocator and Redis uses `jemalloc`, the allocation behavior will be inconsistent.
+
+Secondly, even if Redis always used the system allocator, memory allocated directly by the module would not be visible to Redis: It would not show up in commands such as [`info memory`](https://redis.io/commands/info), and would not be influenced by cleanup operations performed by Redis such as eviction of keys.
```

Comment:
the first word after the colon should be lowercase: "....Redis: it would not show up in commands"